ocaml-lsp-server/src/custom_requests/req_type_enclosing.ml

```diff
@@ -0,0 +1,191 @@
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let of_yojson = function
+  | `Assoc params as full_params ->
+    let verbosity = verbosity_of_yojson params in
+    let range_end = range_end_of_yojson params in
+    let open Option.O in
+    let* index = index_of_yojson params in
+    let text_document_position =
+      Lsp.Types.TextDocumentPositionParams.t_of_yojson full_params
+    in
+    Some { index; range_end; verbosity; text_document_position }
+  | _ -> None
+
+let of_yojson_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match of_yojson params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", expected_params)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
+
+let render_result index result =
+  let typ, enclosings =
+    match result with
+    | None -> ("<no information>", [])
+    | Some (typ, enclosings) ->
+      (typ, List.map ~f:Lsp.Types.Range.yojson_of_t enclosings)
+  in
+  `Assoc
+    [ ("index", `Int index)
+    ; ("enclosings", `List enclosings)
+    ; ("type", `String typ)
+    ]
+
+let config_with_given_verbosity config verbosity =
+  let open Mconfig in
+  { config with query = { config.query with verbosity } }
+
+let with_pipeline state uri verbosity with_pipeline =
+  let doc = Document_store.get state.State.store uri in
+  match Document.kind doc with
+  | `Other -> Fiber.return None
+  | `Merlin merlin ->
+    let open Fiber.O in
+    let* config = Document.Merlin.mconfig merlin in
+    Document.Merlin.with_configurable_pipeline_exn
+      ~config:(config_with_given_verbosity config verbosity)
+      merlin
+      with_pipeline
+
+let make_enclosing_command position index =
+  Query_protocol.Type_enclosing (None, position, Some index)
+
+let get_first_enclosing_index range_end enclosings =
+  let rec aux i = function
+    | (loc, _, _) :: xs -> (
+      let range = Range.of_loc loc in
+      match
+        ( Position.compare range_end range.start
+        , Position.compare range_end range.end_ )
+      with
+      | Ordering.(Gt, Gt)
+      | Ordering.(Eq, Lt)
+      | Ordering.(Gt, Eq)
+      | Ordering.(Eq, Eq)
+      | Ordering.(Gt, Lt) -> Some i
+      | _ -> aux (succ i) xs)
+    | _ -> None
+  in
+  aux 0 enclosings
+
+let dispatch_command pipeline command first_index index =
+  let rec aux i acc = function
+    | (_, `String typ, _) :: _ as enclosings when i = index ->
+      Some
+        ( typ
+        , List.map
+            ~f:(fun (loc, _, _) -> Range.of_loc loc)
+            (List.rev_append acc enclosings) )
+    | curr :: enclosings when i >= first_index ->
+      aux (succ i) (curr :: acc) enclosings
+    | _ :: enclosings -> aux (succ i) acc enclosings
+    | [] -> None
+  in
```

Comment:
I think you just need `List.drop first_index` here?

Reply:
`List.drop` does not exists in `Stdune`.

Reply:
Feel free to try base. And if it doesn't exist, writing it into import.ml is an option too.

Reply:
Fixed into https://github.com/ocaml/ocaml-lsp/pull/1304/commits/fc297726568d120fd3a927266fdee1c90b4a8409